controllers/fluxcd/application-controller.go

```diff
@@ -0,0 +1,453 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package fluxcd
+
+import (
+	"context"
+	"fmt"
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	"kubesphere.io/devops/pkg/api/gitops/v1alpha1"
+	helmv2 "kubesphere.io/devops/pkg/external/fluxcd/helm/v2beta1"
+	kusv1 "kubesphere.io/devops/pkg/external/fluxcd/kustomize/v1beta2"
+	sourcev1 "kubesphere.io/devops/pkg/external/fluxcd/source/v1beta2"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+)
+
+//+kubebuilder:rbac:groups=gitops.kubesphere.io,resources=applications,verbs=watch;get;list
+//+kubebuilder:rbac:groups="",resources=events,verbs=create;patch
+//+kubebuilder:rbac:groups="kustomize.toolkit.fluxcd.io",resources=kustomizations,verbs=watch;get;list;create;update;delete
+//+kubebuilder:rbac:groups="helm.toolkit.fluxcd.io",resources=helmreleases,verbs=watch;get;list;create;update;delete
+//+kubebuilder:rbac:groups="source.toolkit.fluxcd.io",resources=helmcharts,verbs=watch;get;list;create
+
+// ApplicationReconciler is the reconciler of the FluxCD HelmRelease and FluxCD Kustomization
+type ApplicationReconciler struct {
+	client.Client
+	log      logr.Logger
+	recorder record.EventRecorder
+}
+
+// Reconcile sync the Application with underlying FluxCD HelmRelease and FluxCD Kustomization CRD
+func (r *ApplicationReconciler) Reconcile(ctx context.Context, req ctrl.Request) (result ctrl.Result, err error) {
+	r.log.Info(fmt.Sprintf("start to reconcile application: %s", req.String()))
+
+	app := &v1alpha1.Application{}
+	if err = r.Get(ctx, req.NamespacedName, app); err != nil {
+		err = client.IgnoreNotFound(err)
+		return
+	}
+
+	if isFluxApp(app) {
+		err = r.reconcileFluxApp(app)
+	}
+	return
+}
+
+func isFluxApp(app *v1alpha1.Application) bool {
+	return app.Spec.Kind == v1alpha1.FluxCD && app.Spec.FluxApp != nil
+}
+
+func (r *ApplicationReconciler) reconcileFluxApp(app *v1alpha1.Application) (err error) {
+	var at AppType
+	if at, err = isHelmOrKustomize(app); err != nil {
+		return err
+	}
+	if at == HelmRelease {
+		return r.reconcileHelm(app)
+	}
+	if at == Kustomization {
+		return r.reconcileKustomization(app)
+	}
+	return
+}
+
+func (r *ApplicationReconciler) reconcileHelm(app *v1alpha1.Application) (err error) {
+	ctx := context.Background()
```

Comment:
Please use the context from https://github.com/kubesphere/ks-devops/pull/743/files#diff-5da9e5f79765c711202e5d041f78644d98f9502949c1102beebcfbed92c73ad9R51